polyfill/index.d.ts

```diff
@@ -0,0 +1,385 @@
+     * offset changes like Daylight Saving time (DST) transitions.
+     *
+     * This option is only relevant if a `DateTime` value does not exist in the
+     * destination time zone (e.g. near "Spring Forward" DST transitions), or
+     * exists more than once (e.g. near "Fall Back" DST transitions).
+     *
+     * In case of ambiguous or non-existent times, this option controls what
+     * absolute time to return:
+     * - `'earlier'`: The earlier time of two possible times
+     * - `'later'`: The later of two possible times
+     * - `'reject'`: Throw a RangeError instead.
+     *
+     * The default is `'earlier'`.
+     *
+     * Compatibility Note: the legacy `Date` object (and libraries like
+     * moment.js and Luxon) gives the same result as `earlier` when turning the
+     * clock back, and `later` when setting the clock forward.
+     *
+     * */
+    disambiguation: 'earlier' | 'later' | 'reject';
+    // TODO: should there be a `compatible` (to legacy `Date`, moment, etc.)
+    //       disambiguation option? See #614.
+  };
+
+  /**
+   * Options for arithmetic operations like `plus()` and `minus()`
+   * */
+  export type ArithmeticOptions = {
+    /**
+     * Controls handling of out-of-range arithmetic results.
+     *
+     * If a result is out of range, then `'constrain'` will clamp the result to
+     * the allowed range, while `'reject'` will throw a RangeError.
+     *
+     * The default is `'constrain'`.
+     */
+    disambiguation: 'constrain' | 'reject';
+    // TODO: may change name from `disambiguation` to `overflow`. See #607.
+  };
+
+  /**
+   * Options to control `Duration.prototype.minus()` behavior
+   * */
+  export type DurationMinusOptions = {
+    /**
+     * Controls how to deal with subtractions that result in negative overflows
+     *
+     * In `'balanceConstrain'` mode, negative fields are balanced with the next
+     * highest field so that none of the fields are negative in the result. If
+     * this is not possible, a `RangeError` is thrown.
+     *
+     * In `'balance'` mode, all fields are balanced with the next highest field,
+     * no matter if they are negative or not.
+     *
+     * TODO: add an example of each mode.
+     *
+     * The default is `'balanceConstrain'`.
+     *
+     */
+    disambiguation: 'balanceConstrain' | 'balance';
+  };
+
+  export interface DifferenceOptions<T extends string> {
+    /**
+     * The largest unit to allow in the resulting `Temporal.Duration` object.
+     *
+     * Valid values may include `'years'`, `'months'`, `'days'`, `'hours'`,
+     * `'minutes'`, and `'seconds'`, although some types may throw an exception
+     * if a value is used that would produce an invalid result. For example,
+     * `hours` is not accepted by `Date.prototype.difference()`.
+     *
+     * The default depends on the type being used.
+     *
+     * TODO: document which types use which defaults.
+     */
+    largestUnit: T;
+  }
```

Comment:
oops, I understand now what it means.

Reply:
Yep, what's especially nice about using string literal union types in cases like this is the IDE experience where you get autocomplete for the string values. For example, `{disambiguation: 'earlier'}` only requires 6 keystrokes, including the brackets. You also get compiler errors if you either make a typo in a literal string value or if you supply a literal string that's not allowed in that context, e.g.` 'day'` for `YearMonth`.  

We don't have any types (yet) where it would help, but one other nice thing about using literals like this is you can encode incompatible options in the type system, e.g. if `{foo: 'x'; bar: 'y'}` is illegal, you can write a type that shows compiler errors if those options are used together.